File: packages/cpp/tests/test_artifact_signatures.cpp
#include <handoffkit/csp/artifact_security.hpp>

#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <functional>
#include <iostream>
#include <iterator>
#include <string>

#if defined(HANDOFFKIT_WITH_CRYPTO)
#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/pem.h>
#endif

#define REQUIRE(condition)                                                        \
    do {                                                                          \
        if (!(condition)) {                                                       \
            std::cerr << "requirement failed at line " << __LINE__ << ": "       \
                      << #condition << '\n';                                      \
            abort();                                                              \
        }                                                                         \
    } while (false)

namespace {

std::string read_file(const std::filesystem::path& path) {
    std::ifstream input(path, std::ios::binary);
    REQUIRE(input.good());
    return {std::istreambuf_iterator<char>(input), std::istreambuf_iterator<char>()};
}

std::string error_code(const std::function<void()>& operation) {
    try {
        operation();
    } catch (const handoffkit::csp::SecurityError& error) {
        return error.code();
    }
    return {};
}

#if defined(HANDOFFKIT_WITH_CRYPTO)
struct EphemeralKeyPair {
    std::string private_key_pem;
    std::string public_key_pem;
};

std::string bio_value(BIO* bio) {
    char* data = nullptr;
    const auto size = BIO_get_mem_data(bio, &data);
    REQUIRE(size > 0 && data != nullptr);
    return {data, static_cast<std::size_t>(size)};
}

EphemeralKeyPair generate_ed25519_key_pair() {
    EVP_PKEY_CTX* context = EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, nullptr);
    REQUIRE(context != nullptr);
    REQUIRE(EVP_PKEY_keygen_init(context) == 1);
    EVP_PKEY* key = nullptr;
    REQUIRE(EVP_PKEY_keygen(context, &key) == 1);
    EVP_PKEY_CTX_free(context);

    BIO* private_bio = BIO_new(BIO_s_mem());
    BIO* public_bio = BIO_new(BIO_s_mem());
    REQUIRE(private_bio != nullptr && public_bio != nullptr);
    REQUIRE(PEM_write_bio_PrivateKey(private_bio, key, nullptr, nullptr, 0, nullptr, nullptr) == 1);
    REQUIRE(PEM_write_bio_PUBKEY(public_bio, key) == 1);
    EphemeralKeyPair result{bio_value(private_bio), bio_value(public_bio)};
    BIO_free(private_bio);
    BIO_free(public_bio);
    EVP_PKEY_free(key);
    return result;
}
#endif

}  // namespace

int main() {
    using namespace handoffkit::csp;
#if !defined(HANDOFFKIT_WITH_CRYPTO)
    REQUIRE(!artifact_signature_provider_available());
    REQUIRE(error_code([] {
                static_cast<void>(artifact_public_key_fingerprint("not-a-key"));
            }) == "artifact_signature_provider_unavailable");
    return 0;
#else
    REQUIRE(artifact_signature_provider_available());
#ifndef HANDOFFKIT_CONTRACTS_DIR
#error HANDOFFKIT_CONTRACTS_DIR is required for artifact signature conformance
#endif
    const auto vector_path = std::filesystem::path(HANDOFFKIT_CONTRACTS_DIR) /
                             "test-fixtures" / "artifact-signing" / "vector.json";
    const auto vector = nlohmann::json::parse(read_file(vector_path));
    const auto expected = SignedArtifact::from_json(vector.at("signed_artifact"));
    const std::string data = "handoffkit signed artifact\n";
    const auto vector_public = vector.at("public_key_pem").get<std::string>();
    const ArtifactSigningCredential vector_credential{
        expected.signer_identity, vector_public, 1700000000, 1900000000, false};
    ArtifactTrustPolicy vector_policy({vector_credential});
    verify_signed_artifact(data, expected, vector_policy, 1800000000);

    const auto producer = generate_ed25519_key_pair();
    const auto wrong = generate_ed25519_key_pair();
    ArtifactSigner signer(producer.private_key_pem, expected.signer_identity);
    const auto signed_value = signer.sign("artifact-ephemeral", data, expected.created_at);
    const ArtifactSigningCredential trusted{
        expected.signer_identity, producer.public_key_pem, 1700000000, 1900000000, false};
    ArtifactTrustPolicy policy({trusted});
    verify_signed_artifact(data, signed_value, policy, 1800000000);

    REQUIRE(error_code([&] {
                verify_signed_artifact("tampered", signed_value, policy, 1800000000);
            }) == "artifact_integrity_mismatch");

    auto bad_signature = signed_value;
    bad_signature.signature[0] = bad_signature.signature[0] == 'A' ? 'B' : 'A';
    REQUIRE(error_code([&] {
                verify_signed_artifact(data, bad_signature, policy, 1800000000);
            }) == "artifact_signature_invalid");

    auto wrong_identity = signed_value;
    wrong_identity.signer_identity = "spiffe://handoffkit.internal/producer/impostor";
    REQUIRE(error_code([&] {
                verify_signed_artifact(data, wrong_identity, policy, 1800000000);
            }) == "artifact_signer_mismatch");

    ArtifactTrustPolicy wrong_key_policy({ArtifactSigningCredential{
        expected.signer_identity, wrong.public_key_pem, 1700000000, 1900000000, false}});
    REQUIRE(error_code([&] {
                verify_signed_artifact(data, signed_value, wrong_key_policy, 1800000000);
            }) == "artifact_signer_untrusted");

    ArtifactTrustPolicy expired_policy({ArtifactSigningCredential{
        expected.signer_identity, producer.public_key_pem, 1700000000, 1750000000, false}});
    REQUIRE(error_code([&] {
                verify_signed_artifact(data, signed_value, expired_policy, 1800000000);
            }) == "artifact_signer_expired");

    ArtifactTrustPolicy revoked_policy({ArtifactSigningCredential{
        expected.signer_identity, producer.public_key_pem, 1700000000, 1900000000, true}});
    REQUIRE(error_code([&] {
                verify_signed_artifact(data, signed_value, revoked_policy, 1800000000);
            }) == "artifact_signer_revoked");

    ArtifactTrustPolicy disallowed_policy({trusted}, {});
    REQUIRE(error_code([&] {
                verify_signed_artifact(data, signed_value, disallowed_policy, 1800000000);
            }) == "artifact_algorithm_unsupported");
    return 0;
#endif
}

File: packages/python/tests/test_artifact_signatures.py
from __future__ import annotations

import base64
import json
from pathlib import Path

import pytest
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

import handoffkit.csp.security as security_module
from handoffkit.csp import (
    ARTIFACT_ALGORITHM_ECDSA_P256_SHA256,
    ArtifactSignatureError,
    ArtifactSigner,
    ArtifactSigningCredential,
    ArtifactTrustPolicy,
    ArtifactVerifier,
    SignedArtifact,
    detect_ecdsa_p256_sha256_support,
)

VECTOR_PATH = (
    Path(__file__).resolve().parents[2]
    / "contracts"
    / "test-fixtures"
    / "artifact-signing"
    / "vector.json"
)
NOW = 1_800_000_000
IDENTITY = "spiffe://handoffkit.internal/producer/build-1"


@pytest.fixture
def signing_keys() -> dict[str, Ed25519PrivateKey]:
    return {
        "producer": Ed25519PrivateKey.generate(),
        "wrong_signer": Ed25519PrivateKey.generate(),
    }


def signer(
    signing_keys: dict[str, Ed25519PrivateKey],
    name: str = "producer",
    identity: str = IDENTITY,
) -> ArtifactSigner:
    return ArtifactSigner(signing_keys[name], identity)


def credential(
    signing_keys: dict[str, Ed25519PrivateKey],
    name: str = "producer",
    identity: str = IDENTITY,
    *,
    valid_from: int = NOW - 100,
    valid_until: int = NOW + 100,
    revoked: bool = False,
) -> ArtifactSigningCredential:
    public_key_pem = (
        signing_keys[name]
        .public_key()
        .public_bytes(
            serialization.Encoding.PEM,
            serialization.PublicFormat.SubjectPublicKeyInfo,
        )
    )
    return ArtifactSigningCredential(
        signer_identity=identity,
        public_key_pem=public_key_pem,
        valid_from=valid_from,
        valid_until=valid_until,
        revoked=revoked,
    )


def test_ed25519_shared_vector_verifies_without_private_fixture() -> None:
    vector = json.loads(VECTOR_PATH.read_text(encoding="utf-8"))
    data = base64.b64decode(vector["data_base64"], validate=True)
    signed = SignedArtifact.from_dict(vector["signed_artifact"])
    assert signed.canonical_payload().decode("utf-8") == vector["canonical_payload"]
    trusted = ArtifactSigningCredential(
        signer_identity=signed.signer_identity,
        public_key_pem=vector["public_key_pem"],
        valid_from=NOW - 100,
        valid_until=NOW + 100,
    )
    assert ArtifactVerifier.verify_signed_artifact(
        data,
        signed,
        ArtifactTrustPolicy([trusted]),
        now=NOW,
    )


def test_ed25519_artifact_sign_verify_and_canonical_payload(
    signing_keys: dict[str, Ed25519PrivateKey],
) -> None:
    data = b"handoffkit signed artifact\n"
    signed = signer(signing_keys).sign("artifact-1", data, created_at=NOW)
    policy = ArtifactTrustPolicy([credential(signing_keys)])

    canonical = json.loads(signed.canonical_payload())
    assert canonical == {
        "algorithm": "ed25519",
        "artifact_id": "artifact-1",
        "content_hash": "8416cac54bfdbe4faec6d73fdb57ae7cfa81703b311b66de3639e826a185f1e4",
        "created_at": NOW,
        "key_fingerprint": signed.key_fingerprint,
        "signer_identity": IDENTITY,
    }
    assert ArtifactVerifier.verify_signed_artifact(data, signed, policy, now=NOW)


@pytest.mark.parametrize(
    ("mutation", "expected_code"),
    [
        ("content", "artifact_integrity_mismatch"),
        ("signature", "artifact_signature_invalid"),
        ("identity", "artifact_signer_mismatch"),
    ],
)
def test_artifact_tamper_and_wrong_signer_rejected(
    signing_keys: dict[str, Ed25519PrivateKey],
    mutation: str,
    expected_code: str,
) -> None:
    data = b"signed payload"
    signed = signer(signing_keys).sign("artifact-2", data, created_at=NOW)
    candidate = signed
    candidate_data = data
    if mutation == "content":
        candidate_data = b"tampered payload"
    elif mutation == "signature":
        candidate = SignedArtifact.from_dict({**signed.to_dict(), "signature": "AAAA"})
    else:
        candidate = SignedArtifact.from_dict(
            {**signed.to_dict(), "signer_identity": "spiffe://evil.invalid/producer"}
        )

    with pytest.raises(ArtifactSignatureError) as caught:
        ArtifactVerifier.verify_signed_artifact(
            candidate_data,
            candidate,
            ArtifactTrustPolicy([credential(signing_keys)]),
            now=NOW,
        )
    assert caught.value.code == expected_code


def test_untrusted_expired_and_revoked_signers_rejected(
    signing_keys: dict[str, Ed25519PrivateKey],
) -> None:
    data = b"signed payload"
    signed = signer(signing_keys).sign("artifact-3", data, created_at=NOW)

    for policy, code in [
        (
            ArtifactTrustPolicy([credential(signing_keys, "wrong_signer")]),
            "artifact_signer_untrusted",
        ),
        (
            ArtifactTrustPolicy([credential(signing_keys, valid_until=NOW - 1)]),
            "artifact_signer_expired",
        ),
        (
            ArtifactTrustPolicy([credential(signing_keys, revoked=True)]),
            "artifact_signer_revoked",
        ),
        (
            ArtifactTrustPolicy(
                [credential(signing_keys)],
                allowed_algorithms=(),
            ),
            "artifact_algorithm_unsupported",
        ),
    ]:
        with pytest.raises(ArtifactSignatureError) as caught:
            ArtifactVerifier.verify_signed_artifact(data, signed, policy, now=NOW)
        assert caught.value.code == code


def test_signature_field_without_valid_crypto_is_never_verified(
    signing_keys: dict[str, Ed25519PrivateKey],
) -> None:
    signed = signer(signing_keys).sign("artifact-4", b"payload", created_at=NOW)
    forged = SignedArtifact.from_dict({**signed.to_dict(), "signature": ""})
    with pytest.raises(ArtifactSignatureError) as caught:
        ArtifactVerifier.verify_signed_artifact(
            b"payload",
            forged,
            ArtifactTrustPolicy([credential(signing_keys)]),
            now=NOW,
        )
    assert caught.value.code == "artifact_signature_invalid"


def _ecdsa_credential(
    private_key: ec.EllipticCurvePrivateKey,
    *,
    valid_from: int = NOW - 100,
    valid_until: int = NOW + 100,
    revoked: bool = False,
) -> ArtifactSigningCredential:
    public_key_pem = private_key.public_key().public_bytes(
        serialization.Encoding.PEM,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return ArtifactSigningCredential(
        signer_identity=IDENTITY,
        public_key_pem=public_key_pem,
        valid_from=valid_from,
        valid_until=valid_until,
        revoked=revoked,
    )


def test_ecdsa_p256_artifact_sign_verify_and_policy_failures() -> None:
    if not detect_ecdsa_p256_sha256_support():
        pytest.skip("cryptography provider does not expose ECDSA-P256")

    data = b"handoffkit ecdsa artifact\n"
    producer = ec.generate_private_key(ec.SECP256R1())
    wrong_signer = ec.generate_private_key(ec.SECP256R1())
    signed = ArtifactSigner(producer, IDENTITY).sign("artifact-ecdsa", data, created_at=NOW)
    assert signed.algorithm == ARTIFACT_ALGORITHM_ECDSA_P256_SHA256
    policy = ArtifactTrustPolicy(
        [_ecdsa_credential(producer)],
        allowed_algorithms=(ARTIFACT_ALGORITHM_ECDSA_P256_SHA256,),
    )
    assert ArtifactVerifier.verify_signed_artifact(data, signed, policy, now=NOW)

    roundtrip = SignedArtifact.from_dict(signed.to_dict())
    assert ArtifactVerifier.verify_signed_artifact(data, roundtrip, policy, now=NOW)

    for candidate_data, candidate, candidate_policy, expected_code in [
        (b"tampered", signed, policy, "artifact_integrity_mismatch"),
        (
            data,
            SignedArtifact.from_dict({**signed.to_dict(), "signature": "AAAA"}),
            policy,
            "artifact_signature_invalid",
        ),
        (
            data,
            signed,
            ArtifactTrustPolicy(
                [_ecdsa_credential(wrong_signer)],
                allowed_algorithms=(ARTIFACT_ALGORITHM_ECDSA_P256_SHA256,),
            ),
            "artifact_signer_untrusted",
        ),
        (
            data,
            signed,
            ArtifactTrustPolicy(
                [_ecdsa_credential(producer, valid_until=NOW - 1)],
                allowed_algorithms=(ARTIFACT_ALGORITHM_ECDSA_P256_SHA256,),
            ),
            "artifact_signer_expired",
        ),
        (
            data,
            signed,
            ArtifactTrustPolicy(
                [_ecdsa_credential(producer, revoked=True)],
                allowed_algorithms=(ARTIFACT_ALGORITHM_ECDSA_P256_SHA256,),
            ),
            "artifact_signer_revoked",
        ),
    ]:
        with pytest.raises(ArtifactSignatureError) as caught:
            ArtifactVerifier.verify_signed_artifact(
                candidate_data, candidate, candidate_policy, now=NOW
            )
        assert caught.value.code == expected_code


def test_ecdsa_provider_unavailable_fails_closed(monkeypatch: pytest.MonkeyPatch) -> None:
    producer = ec.generate_private_key(ec.SECP256R1())
    monkeypatch.setattr(security_module, "detect_ecdsa_p256_sha256_support", lambda: False)
    with pytest.raises(ArtifactSignatureError) as caught:
        ArtifactSigner(producer, IDENTITY)
    assert caught.value.code == "artifact_algorithm_unsupported"
